.ado/templates/cache-nuget-restore.yml

```diff
@@ -0,0 +1,76 @@
+parameters:
+  # Path to NuGet.config used to restore packages
+  - name: nugetConfigPath
+    type: string
+
+  # Path to the .sln file to restore
+  - name: solutionPath
+    type: string
+
+  # Path to PackageReference global NuGet cache
+  - name: globalPackagesFolder 
+    type: string
+    default: $(Agent.TempDirectory)/.nuget2/packages
+
+steps:
+  - powershell: |
+      nuget locals global-packages -clear
+      Write-Host "##vso[task.setvariable variable=NUGET_PACKAGES]${{ parameters.globalPackagesFolder }}"
+      md $(Agent.TempDirectory)\nupkg
+    displayName: Set isolated global NuGet directory
+
+  - task: Cache@2
+    displayName: Cache global nupkgs
+    inputs:
+      # Cache the packages downloaded to system global cache
+      path: $(Agent.TempDirectory)\nupkg
+      
+      # Cause a cache miss if:
+      #   1. NuGet.config changes
+      #   2. Any changes to a `packages.config` file in the repo or npm depdency
+      #   3. Any changes to a `packages.lock.json` file in the repo or npm depdency (for PackageReference)
+      #
+      # This is scoped to the entire repo + node_modules because we cannot
+      # reason about the dependency graph between packages. It can be keyed
+      # solely to solution lockfile once we are fully PackageReference.
+      key: >
+        "cache-nuget-global.rev3"
+        | "${{ parameters.solutionPath }}"
+        | ${{ parameters.nugetConfigPath }}
+        | **/packages.config
+        | **/packages.lock.json
+
+      # On cache miss, start from another copy for the same solution.
+      # Invalidate the cache if NuGet config changes.
+      restoreKeys: >
+        "cache-nuget-global.rev3"
+        | "${{ parameters.solutionPath }}"
+        | ${{ parameters.nugetConfigPath }}
+
+      cacheHitVar: nugetCacheHit
+
+  - powershell: |
+      $size = @{label="Size";expression={"$([math]::ceiling($_.length / 1mb))MB"}}
+      gci $(Agent.TempDirectory)\nupkg -Recurse -Filter *.nupkg | Sort-Object Length -Descending | Select-Object -First 50 name, $size
+    displayName: Show largest packages
+    condition: ne(variables.nugetCacheHit, 'false')
+
+  - pwsh: |
+      gci -File -Recurse -Filter *.nupkg $(Agent.TempDirectory)\nupkg | ForEach-Object -Parallel {
+        $outputFile = $_.FullName.Replace('$(Agent.TempDirectory)\nupkg','${{ parameters.globalPackagesFolder }}')
+        Copy-Item $_.FullName _Destionation $outputFile -Recurse
+
+        # 7-zip is orders of magnitude faster than Expand-Archive;
+        $outputDir = $_.Directory.FullName.Replace('$(Agent.TempDirectory)\nupkg','${{ parameters.globalPackagesFolder }}')
+        & $env:ProgramFiles\7-Zip\7z.exe x -y "-o$($outputDir)\" "$($_.FullName)"
```

Comment:
Not sure if 7zip does the 'right' thing...
Nuget is a zip-file but it is a zip-file with some semantics and path translations that are supposed to happen.

if the .zip file contains `lib/netcore50/my%20lib.dll`
The extracted file should be `lib/netcore50/my lib.dll`. If 7zip does the right thing, then no worries, else we might have to use a different library and manually extract the package, or actually use nuget.exe for that...

Reply:
and you take care of that later... Please ignore my ignorance :P